tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c

```diff
@@ -0,0 +1,457 @@
+
+static void alt_thread_entry(void)
+{
+	zassert_true(switch_flag == 0,
+		"Alternative thread: switch flag not zero on thread entry\n");
+
+	/* Increment switch flag */
+	switch_flag = 1;
+
+#if defined(CONFIG_NO_OPTIMIZATIONS)
+	zassert_true(p_ztest_thread->arch.basepri == 0,
+		"ztest thread basepri not preserved in swap-out\n");
+#else
+	/* Verify that the main test thread has an initial value of zero
+	 * for state variable thread.arch.basepri.
+	 */
+	zassert_true(p_ztest_thread->arch.basepri == BASEPRI_MODIFIED_1,
+		"ztest thread basepri not preserved in swap-out\n");
+
+	/* Verify original swap return value (set by __swap() */
+	zassert_true(p_ztest_thread->arch.swap_return_value == -EAGAIN,
+		"ztest thread swap-return-value not preserved in swap-out\n");
+#endif
+
+	/* Verify that the main test thread (ztest) has stored the callee-saved
+	 * registers properly in its corresponding callee-saved container.
+	 */
+	verify_callee_saved(
+		(const _callee_saved_t *)&p_ztest_thread->callee_saved,
+		&ztest_thread_callee_saved_regs_container);
+
+	/* Zero the container of the callee-saved registers, to validate,
+	 * later, that it is populated properly.
+	 */
+	memset(&ztest_thread_callee_saved_regs_container,
+		0, sizeof(_callee_saved_t));
+
+#if defined(CONFIG_FLOAT) && defined(CONFIG_FP_SHARING)
+	zassert_true((p_ztest_thread->arch.mode & CONTROL_FPCA_Msk) != 0,
+		"ztest thread mode FPCA flag not updated at swap-out: 0x%0x\n",
+		p_ztest_thread->arch.mode);
+
+	/* Verify that the main test thread (ztest) has stored the FP
+	 *  callee-saved registers properly in its corresponding FP
+	 *  callee-saved container.
+	 */
+	verify_fp_callee_saved((const struct _preempt_float *)
+		&p_ztest_thread->arch.preempt_float,
+		&ztest_thread_fp_callee_saved_regs);
+
+	/* Zero the container of the FP callee-saved registers, to validate,
+	 * later, that it is populated properly.
+	 */
+	memset(&ztest_thread_fp_callee_saved_regs,
+		0, sizeof(_callee_saved_t));
+
+#endif /* CONFIG_FLOAT && CONFIG_FP_SHARING */
+
+	/* Modify the arch.basepri flag of the main test thread, to verify,
+	 * later, that this is passed properly to the BASEPRI.
+	 */
+	p_ztest_thread->arch.basepri = BASEPRI_MODIFIED_2;
+
+#if !defined(CONFIG_NO_OPTIMIZATIONS)
+	/* Modify the arch.swap_return_value flag of the main test thread,
+	 * to verify later, that this value is properly returned by swap.
+	 */
+	p_ztest_thread->arch.swap_return_value = SWAP_RETVAL;
+#endif
+
+	z_move_thread_to_end_of_prio_q(_current);
+
+	/* Zero the callee-saved registers, to simulate that the alternative
+	 * thread is modifying them.
+	 */
+	__asm__ volatile (
+		"mov r0, r7;\n\t"
+		"ldmia %0, {r4-r11};\n\t"
+		"mov r7, r0;\n\t"
+		: : "r" (&ztest_thread_callee_saved_regs_container)
+		: "memory"
+	);
+
+	/* Manually trigger a context-switch, to swap-out
+	 * the alternative test thread.
+	 */
+	__DMB();
+	SCB->ICSR |= SCB_ICSR_PENDSVSET_Msk;
+	__DSB();
+	__ISB();
+}
```

Comment:
What will happen if this thread will be swapped-in back?

Reply:
This is a controlled experiment, so it should not happen with the current code.

Now, in general, I don't think the code paths is any different in case the same thread is -out and -in again. The same code lines are covered, IMHO...
We still want to verify that the BasePri, Ret-Val, callee-saved regs, etc. are preserved...

Reply:
So maybe we should add assert to check if the experiment is really controlled?

Reply:
It is "there",  @pizi-nordic , that's why we have the switch_flag. If it is set to true, means the alternative thread has switched in

Reply:
that's for the main test thread, of course.
But I can add an ASSERT that this thread is not switched-in again, before the main test thread has a chance to resume the experiment.

Thanks, will add.

Reply:
Done